[FEATURE] Générer les tests de certif à partir des acquis validés directs (et plus des épreuves bien répondues) (PIX-968)

sroccaserra opened the pull request:
## ☕️ Contexte

Analyse démarrée dans le cadre de l’internationalisation, plusieurs acquis ont été remplacés d’un coup (voir https://1024pix.atlassian.net/browse/PIX-967 ).

## 🦄 Problème

Aujourd'hui, l'algorithme qui génère les tests de certification sélectionne les acquis à certifier en partant des épreuves réussies par le candidat. Problème, quand une épreuve réussie par le candidat a été périmée entre son positionnement et son test de certification, l'acquis (les acquis) correspondants est (sont) écarté(s) des acquis à certifier. C'est dommage, car souvent il existe d'autres épreuves non périmées qui valident le même acquis.

Il y a donc un risque d’avoir des certifications qui tombent en “error” car pas d’assez d'épreuves à poser au candidat pendant son test.

Autre problème, les épreuves possibles pour valider un acquis sont identifiées par le nom d'acquis. Ce n'est pas ce qu'on veut, on veut passer par l'identifiant de l'acquis.

Autre problème : l'algorithme actuel passe par les _knowledge elements_ deux fois, une fois pour la génération du profil de positionnement du candidat, et une deuxième fois pour la sélection des épreuves du test de certification. C'est compliqué.

## 🤖 Solution

Simplifier l'algorithme en sélectionnant les acquis à valider lors de la génération du profil de positionnement, et les passer à l'algo de choix des épreuves. Ça corrige aussi le problème des épreuves périmées entre le positionnement et la certification.

Corriger également l'identification des épreuves qui valident un acquis en utilisant l'id de l'acquis au lieu de son nom.

## 👓 Remarque

Étape d'après facile qu'on a en tête : transférer à l'appelant la responsabilité de récupérer les id d'épreuves bien répondues et les passer en arguments au `pickCertificationChallenges()`.

## 📈 Stats

### Stats sur 100 utilisateurs

Les stats de simulation de génération de tests sur 100 utilisateurs récents en terme de niveaux des épreuves par compétences :

- 82 tests ont des niveaux d’épreuves identiques
- 13 tests ont un ou plusieurs niveaux d’épreuve supérieurs sur une compétence (dont un test qui a une compétence certifiable en plus)
- 5 tests ont un niveau d’épreuve en moins sur une compétence (voir analyse ci-dessous)

Les 18 différences pour 100 utilisateurs :
```
recDH19F7kKrfL3Ii       5:5:5 6:5:5
recMiZPNl7V1hyE1d       5:4:2 5:4:4
recFpYXCKcyhLI3Nu       3:1:1 3:3:2
recgxqQfz3BqEbtzh       4:4:3 4:4:4
recIkYm646lrGvLNT       6:5:5 6:6:5
recOdC9UDVJbAXHAm       4:3:3 5:4:3
recbDTF8KwupqkeZ6       6:5:5 6:6:5
recgxqQfz3BqEbtzh       4:4:3 5:4:4
recMiZPNl7V1hyE1d       5:5:4 5:4:4  <--
recOdC9UDVJbAXHAm       5:4:4 4:4:4  <-- Ces 5 tests de certif sur 100
recOdC9UDVJbAXHAm       5:5:4 5:4:4  <-- semblent avoir une compétence
recOdC9UDVJbAXHAm       5:4:4 4:4:4  <-- moins bien testée
recOdC9UDVJbAXHAm       5:5:4 5:4:4  <--
recbDTF8KwupqkeZ6       4:2   4:4:2
recIhdrmCuEmCDAzj       5:5:5 6:5:5
recDH19F7kKrfL3Ii       5:5:5 6:5:5
recbDTF8KwupqkeZ6       6:5:5 6:6:5
recfr0ax8XrfvJ3ER       6:5:5 6:6:5
```

#### 4 Cas de dégradation de la compétence `recOdC9UDVJbAXHAm`

Pour ces 4 cas je trouve le même problème.

En suivant le déroulé de la version ancienne, je trouve :
- une épreuve déjà répondue qui valide les acquis `@outilsDiaporama3` et `@outilsDiaporama5`
- un ke uniquement pour `@outilsDiaporama3` 🤔
- pourtant c'est une épreuve sur `@outilsDiaporama5` qui serait choisie dans la version ancienne 🤔

Hypothèse : l'acquis `@outilsDiaporama5` s'est ajouté à l'épreuve après le positionnement du candidat ? Dans ce cas il semble logique de chercher un autre acquis. Si cette hypothèse est juste, la nouvelle version est meilleure.

#### Cas de dégradation de la compétence `recMiZPNl7V1hyE1d`

Même topo, en suivant le déroulé de la version ancienne, je trouve :
- une épreuve déjà répondue qui valide les acquis `@outilscollaboratifs4` et `@outilscollaboratifs5`
- un ke uniquement pour `@outilscollaboratifs4` 🤔
- pourtant c'est une épreuve sur `@outilscollaboratifs5` qui serait choisie dans la version ancienne 🤔

Pour les mêmes raisons, la nouvelle version semble meilleure.

### Stats sur 1000 utilisateurs

Sur 1000 utilisateurs, soit 10217 compétences, on trouve :

- 9333 compétences (91.5 %) identiques en terme de niveaux d'acquis évalués
- 835 compétences (8 %) qui sont mieux évaluées sur la branche (soit acquis de niveaux supérieur évalués, soit une épreuve en plus)
- 49 compétences (0.5 %) qui sont moins bien évaluées sur la branche

Note : j'ai utilisé `diff -y --suppress-common-lines` et ce script awk pour faire les stats :
```awk
BEGIN { total=0 ; ok=0 ; ko=0 ; new=0 ; lost=0 }
END { print "total: " total ", ok: " ok ", ko: " ko ", new: " new  ", lost: " lost}

// { ++total }
/^[0-9]/ { if ($3 <= $7) { ok++ ; print "ok: " $0 } else { ko++ ; print "ko: " $0 } }
/[>]/ { ++new ; print "new: " $0 }
/[<]/ { ++lost ; print "lost: " $0 }
```

sroccaserra committed:
Put dir. valid. skills in placementProfile

For certification test generation only.

And certification challenge selection is now done using direct validated
skills instead of skills derived from already answered challenges.

pix-service commented:
I'm deploying this PR to these urls:

- App: https://app-pr1860.review.pix.fr
- Orga: https://orga-pr1860.review.pix.fr
- Certif: https://certif-pr1860.review.pix.fr
- Admin: https://admin-pr1860.review.pix.fr
- API: https://api-pr1860.review.pix.fr/api/

 Please check it out!

jonathanperret commented on api/lib/domain/services/certification-challenges-service.js:
```diff
@@ -8,9 +8,21 @@ const KnowledgeElement = require('../models/KnowledgeElement');
+    const associatedKnowledgeElements = knowledgeElements.filter((knowledgeElement) => knowledgeElement.answerId === answer.id);
+    return {
+      challengeId: answer.challengeId,
+      competenceId: associatedKnowledgeElements[0].competenceId, // Ici on doit être confiant dans le fait qu'une answer lié à plusieurs KE (QROCmDep) soit liés à des skills de la même compétence
```
C'est plutôt le challenge qui est lié à des skills non ?
```suggestion
      competenceId: associatedKnowledgeElements[0].competenceId, // Ici on doit être confiant dans le fait qu'un challenge lié à plusieurs skills (QROCmDep) soit liée à des skills de la même compétence
```

sroccaserra replied:
Oui, c'est un raccourci, peut-être que ça c'est mieux :
```suggestion
      // Ici on doit être confiant dans le fait qu'une réponse liée à
      // plusieurs KEs (QROCmDep) soit liée à une épreuve liée à des acquis
      // de la même compétence
      competenceId: associatedKnowledgeElements[0].competenceId,
```

jonathanperret replied:
Le truc qui m'embêtait au départ c'est qu'une réponse liée à plusieurs KE, c'est la norme à cause de l'inférence ! Bien sûr les KE qui arrivent ici ne sont que les _directs_, mais sans avoir le contexte c'est pas évident de s'en souvenir.

jonathanperret commented on api/tests/integration/domain/services/pick-certification-challenge_test.js:
```diff
@@ -675,6 +675,111 @@ describe('Integration | CertificationChallengeService | pickCertificationChallen
     expect(challenges[0].challengeId).to.equal('recArea1_Competence1_Tube1_Skill1_And_Skill2_Challenge');
   });
+
+  describe('When skills are moved in the referential after placement', () => {
```
```suggestion
  describe('When skills are moved in the learning content repository after placement', () => {
```

sroccaserra committed:
Update script to generate certif tests: use stderr

Redirect irrelvant output to stderr, stdout is reserved for stats only.
Useful to write stats to a file, and diff stats only.

P-Jeremy committed:
Add `list()` method to skill repository

sroccaserra committed:
Add USER_ID param to certif course script

sroccaserra committed:
Match challenge skills by id, instead of name

sroccaserra committed:
Rename userCompetences to use...sWithOrderedSkills

sroccaserra committed:
Filter certifiable userCompetences earlier

Giroteti committed:
Use challengesIds directly

laura-bergoens commented on api/lib/domain/services/placement-profile-service.js:
```diff
@@ -114,6 +133,17 @@ async function getPlacementProfilesWithSnapshotting({ userIdsAndDates, competenc
+function _matchingValidatedDirectSkillsForCompetence(knowledgeElementsForCompetence, skillMap) {
+  const competenceSkills = knowledgeElementsForCompetence.map((ke) => {
+    const skill = skillMap.get(ke.skillId);
+    if (ke.status === KnowledgeElement.StatusType.VALIDATED && ke.source ===  KnowledgeElement.SourceType.DIRECT) {
```
**suggestion (non-blocking) :**
Je propose de déplacer cette évaluation dans une méthode du modèle `KnowledgeElement` pour avoir
```js
if(ke.isDirectlyValidated) {
/*...*/
}
```

sroccaserra replied:
Done, merci

laura-bergoens commented on api/lib/domain/services/placement-profile-service.js:
```diff
@@ -114,6 +133,17 @@ async function getPlacementProfilesWithSnapshotting({ userIdsAndDates, competenc
+    }
+  });
+
+  return _.compact(competenceSkills);
```
**suggestion (non-blocking) :**
En principe il ne devrait pas y en avoir besoin. La méthode du knowledge-element-repository pour obtenir la collection de KE garantit retourner le KE le plus récent **par acquis**.

laura-bergoens commented on api/lib/domain/models/UserCompetence.js:
```diff
@@ -16,6 +16,7 @@ class UserCompetence {
     pixScore,
     estimatedLevel,
     // includes
+    skills = [],
```
**thought (non-blocking) :**
En prescription aussi on utilise le placementProfile pour certaines fonctionnalités. Je sens qu'on a un modèle ici dont l'usage commence à se multiplier et diverger. Car nous, on va toujours avoir une liste vide ici dans notre flow ce qui est un peu bizarre. Je fais la remarque pas tant pour que vous preniez le sujet, mais pour nous. Je pense que c'est à nous de voir si ces modèles là nous conviennent toujours (_duplication over wrong abstraction_ toussa ;) ).

sroccaserra replied:
Oui, on s'est posé la question, le sujet m'intéresse

laura-bergoens replied:
C'est un truc à voir du côté prescription je pense. Je me l'ajoute à ma liste et je te ping alors quand je re-traiterai le sujet de notre côté (je mettrai Arthur dans la boucle aussi, c'est son genre de sujet)

laura-bergoens commented on api/lib/domain/services/certification-challenges-service.js:
```diff
@@ -16,46 +16,33 @@ module.exports = {
       .findUniqByUserIdGroupedByCompetenceId({ userId: placementProfile.userId, limitDate: placementProfile.profileDate });
-
     const knowledgeElements = KnowledgeElement.findDirectlyValidatedFromGroups(knowledgeElementsByCompetence);
     const answerIds = _.map(knowledgeElements, 'answerId');
```
**question (non-blocking) :**
J'ai failli pas comprendre ici car, dans le placementProfile, vous faites justement en sorte de récupérer la liste des acquis à évaluer en certif depuis les directlyValidated KEs, et là vous re-récupérez les KEs et j'ai failli crier au scandale. Mais je comprends qu'en fait c'est juste pour récupérer les `answerIds` c'est ça ?

sroccaserra replied:
Oui, et je dirais même plus : on fait tout ça pour avoir la liste des `alreadyAnsweredChallengeIds`. Étape d'après facile qu'on a en tête : transférer à l'appelant la responsabilité de récupérer les id d'épreuves bien répondues et les passer en arguments au `pickCertificationChallenges()`.

laura-bergoens commented on api/lib/domain/services/certification-challenges-service.js:
```diff
@@ -16,46 +16,33 @@ module.exports = {
+      .map((id) => { return { id }; });
 
+    const allFrFrOperativeChallenges = await challengeRepository.findFrenchFranceOperative();
     const userCompetences = UserCompetence.orderSkillsOfCompetenceByDifficulty(placementProfile.userCompetences);
```
**thought (non-blocking) :**
Un jour ça serait bien de déplacer cette logique dans une méthode du placementProfile peut-être ? qui lui-même parcourt ses userCompetences et leur demande à chacune de faire le tri? . Je comprends pas trop pourquoi c'est une méthode statique.

sroccaserra replied:
Ah oui ça paraît une bonne idée !

laura-bergoens commented on api/lib/domain/services/certification-challenges-service.js:
```diff
@@ -16,46 +16,33 @@ module.exports = {
     userCompetences.forEach((userCompetence) => {
+      if (!userCompetence.isCertifiable()) {
+        return;
+      }
```
**suggestion (non-blocking) :**
On pourrait pas filtrer la bonne sous-collection juste au-dessus pour faire le map comme il faut ?
```js
const certifiableUserCompetences = placementProfile.getCertifiableUserCompetences();
certifiableUserCompetences.forEach(....
```

sroccaserra replied:
Oui, c'est fait dans un commit d'après 👍

laura-bergoens reviewed (approved):
Cette itération allège vraiment bien le service, donc c'est super cool !
En plus, j'ai clairement l'impression (la certitude en fait) que la couverture de tests autour de ce service est de mieux en mieux. Bravo l'équipe !
_One step at a time_

aceol commented on api/scripts/generate-certification-test-statistics.js:
```diff
@@ -7,6 +7,7 @@ const placementProfileService = require('../lib/domain/services/placement-profil
 const certificationChallengeService = require('../lib/domain/services/certification-challenges-service');
 
 const USER_COUNT = ~~process.env.USER_COUNT || 100;
+const USER_ID = ~~process.env.USER_ID || null;
```
je serais pour retirer ce ~~ qui fait bugger mon cerveau à chaque fois que je le vois.
De plus je vois pas son interêt ici

Anne-Gaelle-S replied:
**question (non bloquant)**
il fait quoi ce "~~" ?

laura-bergoens replied:
C'est un "tricks" pour convertir une valeur en integer. Si c'est pas lisible pour le plus grand nombre je vois pas de problème à le remplacer aussi. L'avantage c'est que c'est très concis et que ça ne se soucie pas du type de départ.

laura-bergoens replied:
en fait la règle c'est  :
~N = -(N+1)       sachant qu'au passage ça convertit N en entier 32 bits.
Du coup mettons que par exemple on a process.env.USER_ID = '123' :
(première tilde) --->  ~'123' = -(123+1) = -124
(deuxième tilde) --->  ~(-124) = -(-124 + 1) = 123 TADA :tada:

laura-bergoens replied:
mais j'avoue c'est :imp: quand on le sait pas

aceol replied:
C'est "souvent" utiliser pour transformer un float en entier

sroccaserra replied:
J'ai mis `parseInt()` à la place, dans ce contexte c'est ça qu'on veut 👍

aceol commented on api/lib/domain/services/certification-challenges-service.js:
```diff
@@ -16,46 +16,33 @@ module.exports = {
     userCompetences.forEach((userCompetence) => {
+      if (!userCompetence.isCertifiable()) {
+        return;
+      }
```
@laura-bergoens on pourrait faire ca?
```suggestion
    userCompetences
        .filter(userCompetence => userCompetence.isCertifiable())
        .forEach((userCompetence) => {
```

sroccaserra replied:
Oui, c'est fait dans un commit après

aceol commented on api/lib/domain/services/certification-challenges-service.js:
```diff
@@ -16,46 +16,33 @@ module.exports = {
+          const challengesToValidateCurrentSkill = Challenge.findBySkill({ challenges: allFrFrOperativeChallenges, skill });
+          const challengesLeftToAnswer = _.differenceBy(challengesToValidateCurrentSkill, alreadyAnsweredChallengeIds, 'id');
 
           const challengesPoolToPickChallengeFrom = (_.isEmpty(challengesLeftToAnswer)) ? challengesToValidateCurrentSkill : challengesLeftToAnswer;
```
```suggestion
          const challengesPoolToPickChallengeFrom = _.isEmpty(challengesLeftToAnswer) ? challengesToValidateCurrentSkill : challengesLeftToAnswer;
```

sroccaserra replied:
Fait, merci

aceol commented on api/lib/domain/services/certification-challenges-service.js:
```diff
@@ -16,46 +16,33 @@ module.exports = {
 
           const challengesPoolToPickChallengeFrom = (_.isEmpty(challengesLeftToAnswer)) ? challengesToValidateCurrentSkill : challengesLeftToAnswer;
+          if (_.isEmpty(challengesPoolToPickChallengeFrom)) {
+            return;
```
est-ce un cas normal? Devrais t'on logguer un message ici?

aceol commented on api/lib/domain/services/placement-profile-service.js:
```diff
@@ -114,6 +133,17 @@ async function getPlacementProfilesWithSnapshotting({ userIdsAndDates, competenc
   return placementProfilesList;
 }
 
+function _matchingValidatedDirectSkillsForCompetence(knowledgeElementsForCompetence, skillMap) {
+  const competenceSkills = knowledgeElementsForCompetence.map((ke) => {
+    const skill = skillMap.get(ke.skillId);
+    if (ke.status === KnowledgeElement.StatusType.VALIDATED && ke.source ===  KnowledgeElement.SourceType.DIRECT) {
+      return skill;
+    }
+  });
+
```
```suggestion
function _matchingValidatedDirectSkillsForCompetence(knowledgeElementsForCompetence, skillMap) {
  const competenceSkills = knowledgeElementsForCompetence.filter((ke) => ke.isDirectlyValidated) // comme suggéré par @laura-bergoens 
   // (ke.status === KnowledgeElement.StatusType.VALIDATED && ke.source ===  KnowledgeElement.SourceType.DIRECT) 
.map(ke => skillMap.get(ke.skillId))
  });

```

sroccaserra replied:
Done, merci

aceol commented on api/tests/integration/domain/services/pick-certification-challenge_test.js:
```diff
@@ -675,6 +679,111 @@ describe('Integration | CertificationChallengeService | pickCertificationChallen
+     *
+     * Cet acquis a maintenant été déplacé dans la compétence Pix+ 7.1
+     *
+     * On veut que le test de certfis contienne une épreuve correspondant à
```
```suggestion
     * On veut que le test de certifs contienne une épreuve correspondant à
```

sroccaserra replied:
Merci, et ça me rappelle que ce commentaire n'est pas bon, il faut qu'on le corrige / supprime

Anne-Gaelle-S commented on api/scripts/generate-certification-test-statistics.js:
```diff
@@ -53,18 +53,18 @@ async function _generateCertificationTest(userId, competences) {
 async function main() {
   try {
-    console.log(`Récupération de ${USER_COUNT} utilisateurs certifiables...`);
+    console.error(`Récupération de ${USER_COUNT} utilisateurs certifiables...`);
```
**question (non bloquant)**
Juste par curiosité, pourquoi on met console.error au lieu des console.log ?

laura-bergoens replied:
Je pense que c'est pour avoir plus de contrôle sur la sortie standard utilisée pour afficher le message. En mettant console.error, j'imagine que, lancé dans un terminal, ça sort sur la sortie standard d'erreur stderr. Tu peux donc avoir un contrôle plus fin lorsque tu veux rediriger les logs vers un fichier en ligne de commande.

```sh
# redirige les console.log() (sortie standard stdout) vers un fichier out.txt 
# et les console.err() (stderr) vers un fichier err.txt
node mon-scripts 1> out.txt 2> err.txt
# qu'on peut aussi écrire 
node mon-scripts > out.txt 2> err.txt
```

sroccaserra replied:
Exactement

aceol commented on api/lib/domain/models/Challenge.js:
```diff
@@ -102,7 +102,7 @@ class Challenge {
 
   hasSkill(searchedSkill) {
-    return this.skills.filter((skill) => skill.name === searchedSkill.name).length > 0;
+    return this.skills.filter((skill) => skill.id === searchedSkill.id).length > 0;
```
return this.skills.some((skill) => skill.id === searchedSkill.id);

sroccaserra replied:
Fait, merci

Giroteti committed:
Remove unused repository stub in test

Anne-Gaelle-S commented on api/lib/domain/services/placement-profile-service.js:
```diff
@@ -114,6 +133,17 @@ async function getPlacementProfilesWithSnapshotting({ userIdsAndDates, competenc
 }
 
+function _matchingValidatedDirectSkillsForCompetence(knowledgeElementsForCompetence, skillMap) {
+  const competenceSkills = knowledgeElementsForCompetence.map((ke) => {
+    const skill = skillMap.get(ke.skillId);
+    if (ke.status === KnowledgeElement.StatusType.VALIDATED && ke.source ===  KnowledgeElement.SourceType.DIRECT) {
+      return skill;
+    }
+  });
```
**suggestion (non bloquant)**
J'ai un peu de mal à comprendre ce qu'on cherche à faire dans cette fonction... 
On essaie de récupérer les skills dont les KE sont directs + validés ?
Si oui je me demande si on ne pourrait pas renommer la fonction pour qu'on puisse comprendre plus vite ce qu'elle fait. Ex: `getSkillsFromValidatedDirectKe` ? Ou autre ?

`:EDIT:` ce que propose Antoine en dessous me paraît plus simple à comprendre

sroccaserra replied:
C'est fait, merci

Anne-Gaelle-S commented on api/lib/infrastructure/repositories/skill-repository.js:
```diff
@@ -14,6 +14,11 @@ function _toDomain(skillData) {
 
 module.exports = {
 
+  async list() {
```
**question (non bloquant)**
Pour ma culture, c'est quoi la différence entre `findActiveByCompetenceId`, `findOperativeByCompetenceId` et `list` ? (en dehors du fait qu'on recherche par Id les 2 premier)

laura-bergoens replied:
C'est lié à la problématique qu'on a eu y'a quelques mois (tu sais préservons nos campagnes, préservons la certif, etc...).
En gros, concernant le choix des acquis à évaluer selon les différents contextes :
- Evaluation autonome : on ne prend **QUE** les acquis actifs
- Campagne : on prend les acquis actifs **ET** les acquis archivés 
- Certif : on prend les acquis actifs **ET** les acquis archivés 

_Operative_ est le terme qui a été choisi pour signifier actif ou archivé

Anne-Gaelle-S reviewed (approved):
Quelques questions, mais rien de bloquant :)

sroccaserra committed:
Add KnowledgeElement.isDirectlyValidated() method

sroccaserra committed:
Simplify Challenge.hasSkill() method

sroccaserra committed:
Change mysterious `~~` operator in script

Giroteti commented on api/lib/domain/services/placement-profile-service.js:
```diff
@@ -74,7 +73,7 @@ function _createUserCompetencesV2({
     });
 
-    const competenceDirectValidatedSkills = _matchingValidatedDirectSkillsForCompetence(knowledgeElementsForCompetence, skillMap);
+    const competenceDirectValidatedSkills = _matchingDirectlyValidatedSkillsForCompetence(knowledgeElementsForCompetence, skillMap);
```
on peut renommer la variable aussi je pense : competenceDirect**ly**ValidatedSkills

sroccaserra replied:
C'est fait, merci

sroccaserra committed:
Add level stats in certification test stats script
